hal/armv7m/stm32/l4/stm32l4.c

```diff
@@ -0,0 +1,490 @@
+
+	t = *(base + gpio_odr) & ~(!(u32)val << pin);
+	*(base + gpio_odr) = t | !!(u32)val << pin;
+
+	return 0;
+}
+
+
+int _stm32_gpioSetPort(unsigned int d, u16 val)
+{
+	volatile u32 *base;
+
+	if (d > pctl_gpioi)
+		return -1;
+
+	base = stm32_common.gpio[d - pctl_gpioa];
+	*(base + gpio_odr) = val;
+
+	return 0;
+}
+
+
+int _stm32_gpioGet(unsigned int d, u8 pin, u8 *val)
+{
+	volatile u32 *base;
+
+	if (d > pctl_gpioi || pin > 15)
+		return -1;
+
+	base = stm32_common.gpio[d - pctl_gpioa];
+	*val = !!(*(base + gpio_idr) & (1 << pin));
+
+	return 0;
+}
+
+
+int _stm32_gpioGetPort(unsigned int d, u16 *val)
+{
+	volatile u32 *base;
+
+	if (d > pctl_gpioi)
+		return -1;
+
+	base = stm32_common.gpio[d - pctl_gpioa];
+	*val = *(base + gpio_idr);
+
+	return 0;
+}
+
+
+/* Watchdog */
+
+
+void _stm32_wdgReload(void)
+{
+#if defined(WATCHDOG) && defined(NDEBUG)
+	*(stm32_common.iwdg + iwdg_kr) = 0xaaaa;
+#endif
+}
+
+
+void _stm32_init(void)
+{
+	u32 i;
+	static const int gpio2pctl[] = { pctl_gpioa, pctl_gpiob, pctl_gpioc,
+		pctl_gpiod, pctl_gpioe, pctl_gpiof, pctl_gpiog, pctl_gpioh, pctl_gpioi };
+
+	/* Base addresses init */
+	stm32_common.rcc = (void *)0x40021000;
+	stm32_common.pwr = (void *)0x40007000;
+	stm32_common.scb = (void *)0xe000e000;
+	stm32_common.rtc = (void *)0x40002800;
+	stm32_common.syscfg = (void *)0x40010000;
+	stm32_common.iwdg = (void *)0x40003000;
+	stm32_common.gpio[0] = (void *)0x48000000; /* GPIOA */
+	stm32_common.gpio[1] = (void *)0x48000400; /* GPIOB */
+	stm32_common.gpio[2] = (void *)0x48000800; /* GPIOC */
+	stm32_common.gpio[3] = (void *)0x48000c00; /* GPIOD */
+	stm32_common.gpio[4] = (void *)0x48001000; /* GPIOE */
+	stm32_common.gpio[5] = (void *)0x48001400; /* GPIOF */
+	stm32_common.gpio[6] = (void *)0x48001800; /* GPIOG */
+	stm32_common.gpio[7] = (void *)0x48001c00; /* GPIOH */
+	stm32_common.gpio[8] = (void *)0x48002000; /* GPIOI */
+	stm32_common.flash = (void *)0x40022000;
+
+	/* Enable System configuration controller */
+	_stm32_rccSetDevClock(pctl_syscfg, 1);
+
+	/* Enable power module */
+	_stm32_rccSetDevClock(pctl_pwr, 1);
+
+	_stm32_rccSetCPUClock(16 * 1000 * 1000);
+
+	/* Disable all interrupts */
+	*(stm32_common.rcc + rcc_cier) = 0;
+
+	hal_cpuDataMemoryBarrier();
+
+	/* GPIO init */
+	for (i = 0; i < sizeof(stm32_common.gpio) / sizeof(stm32_common.gpio[0]); ++i)
+		_stm32_rccSetDevClock(gpio2pctl[i], 1);
+
+	hal_cpuDataMemoryBarrier();
+
+#if defined(WATCHDOG) && defined(NDEBUG)
+	/* Init watchdog */
+	/* Enable write access to IWDG */
+	*(stm32_common.iwdg + iwdg_kr) = 0x5555;
+
+	/* Set prescaler to 256, ~30s interval */
+	*(stm32_common.iwdg + iwdg_pr) = 0x06;
+	*(stm32_common.iwdg + iwdg_rlr) = 0xfff;
+
+	_stm32_wdgReload();
+
+	/* Enable watchdog */
+	*(stm32_common.iwdg + iwdg_kr) = 0xcccc;
+#endif
+
+#ifdef NDEBUG
+	*(u32 *)0xE0042004 = 0;
+#endif
+
+	/* Enable UsageFault, BusFault and MemManage exceptions */
+	*(stm32_common.scb + scb_shcsr) |= (1 << 16) | (1 << 17) | (1 << 18);
```

Comment:
This line is duplicated:
https://github.com/phoenix-rtos/plo/blob/ab4d7e29a3acef43f4a7375f9b67d414b990b5e3/hal/armv7m/cpu.c#L272

Reply:
Done